pkg/controller/direct/secretmanager/secret_controller.go

```diff
@@ -0,0 +1,367 @@
+		updateMask.Paths = append(updateMask.Paths, "topics")
+	}
+	if !common.DeepEqual_StringAndTimestampPb(*desired.Spec.ExpireTime, a.actual.GetExpireTime()) {
+		updateMask.Paths = append(updateMask.Paths, "expire_time")
+	}
+	if !reflect.DeepEqual(desired.Spec.Rotation.NextRotationTime, a.actual.GetRotation().GetNextRotationTime()) {
+		updateMask.Paths = append(updateMask.Paths, "rotation.next_rotation_time")
+	}
+	if !reflect.DeepEqual(desired.Spec.Rotation.RotationPeriod, a.actual.GetRotation().GetRotationPeriod()) {
+		updateMask.Paths = append(updateMask.Paths, "rotation.rotation_period")
+	}
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id.FullyQualifiedName())
+		return nil
+	}
+
+	req := &secretmanagerpb.UpdateSecretRequest{
+		UpdateMask: updateMask,
+		Secret:     resource,
+	}
+	updated, err := a.gcpClient.UpdateSecret(ctx, req)
+	if err != nil {
+		return fmt.Errorf("Secret %s waiting update: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully updated Secret", "name", a.id.FullyQualifiedName())
+
+	status := &krm.SecretManagerSecretStatus{}
+	status.ObservedState = SecretManagerSecretStatusObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.Name = updated.Name
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.SecretManagerSecret{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(SecretManagerSecretSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *Adapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("deleting Secret", "name", a.id.FullyQualifiedName())
+
+	req := &secretmanagerpb.DeleteSecretRequest{Name: a.id.FullyQualifiedName()}
+	err := a.gcpClient.DeleteSecret(ctx, req)
+	if err != nil {
+		return false, fmt.Errorf("deleting Secret %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully deleted Secret", "name", a.id.FullyQualifiedName())
+	return true, nil
+}
+
+func setStatus(u *unstructured.Unstructured, typedStatus any) error {
+	status, err := runtime.DefaultUnstructuredConverter.ToUnstructured(typedStatus)
+	if err != nil {
+		return fmt.Errorf("error converting status to unstructured: %w", err)
+	}
+
+	old, _, _ := unstructured.NestedMap(u.Object, "status")
+	if old != nil {
+		status["conditions"] = old["conditions"]
+		status["observedGeneration"] = old["observedGeneration"]
+		status["externalRef"] = old["externalRef"]
```

Comment:
Nit: I'm not sure why we are copying externalRef - I don't think we have to...

Reply:
Ah! At the time I added this template, we don't have this wonderful code! https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/d4cc447191eddae1a6def67146e66c0658f283a1/pkg/controller/direct/directbase/operations.go#L95
You are right, we don't need to now

Reply:
Updated the template so future controllers can use opt to update #2919 2919